components/expression/__examples__/base.jsx

```diff
@@ -0,0 +1,273 @@
+import React from 'react';
+import find from 'lodash.find';
+
+// `~` is replaced with design-system-react at runtime
+import IconSettings from '~/components/icon-settings';
+import Expression from '~/components/expression';
+import ExpressionGroup from '~/components/expression/group';
+import ExpressionCondition from '~/components/expression/condition';
+
+const ResourcesList = [
+	{
+		id: '111',
+		label: 'Resource 1',
+	},
+	{
+		id: '112',
+		label: 'Resource 2',
+	},
+	{
+		id: '113',
+		label: 'Resource 3',
+	},
+	{
+		id: '114',
+		label: 'Resource 4',
+	},
+];
+
+const OperatorsList = [
+	{
+		id: '1',
+		label: 'Equals',
+	},
+	{
+		id: '2',
+		label: 'Does Not Equals',
+	},
+	{
+		id: '3',
+		label: 'Greater Than',
+	},
+	{
+		id: '4',
+		label: 'Less Than',
+	},
+];
+
+class Example extends React.Component {
+	static displayName = 'ExpressionBaseExample';
+
+	constructor(props) {
+		super(props);
+		this.state = {
+			conditions: [
+				{
+					isGroup: false,
+					resource: '',
+					operator: '',
+					value: '',
+				},
+			],
+			triggerType: 'all',
+			customLogic: '',
+		};
+	}
+
+	static getTriggerType(i, trigger) {
+		if (trigger === 'custom') return String(i + 1);
+		if (i > 0) {
+			if (trigger === 'all') return 'AND';
+			if (trigger === 'any') return 'OR';
+		}
+		return '';
+	}
+
+	updateData(i, val, type) {
+		const { conditions } = this.state;
+		if (type === 'value') conditions[i].value = val;
+		else conditions[i][type] = val.selection[0].id;
+		this.setState({ conditions });
+	}
+
+	updateSubData(i, j, val, type) {
+		const { conditions } = this.state;
+		if (type === 'value') conditions[i].conditions[j].value = val;
+		else conditions[i].conditions[j][type] = val.selection[0].id;
+		this.setState({ conditions });
+	}
+
+	addCondition() {
+		const { conditions } = this.state;
+		const newCondition = {
+			isGroup: false,
+			resource: '',
+			operator: '',
+			value: '',
+		};
+		conditions.push(newCondition);
```

Comment:
spread `{...}` and `array.concat()` are your friends here. You shouldn't need `update` then. See the tree for some examples. `push` mutates variables which is why `update` is needed here.

Reply:
Another option is to do this:
`const conditions = JSON.parse(JSON.stringify(this.state.conditions));`, which guarantees conditions is a new array. Note that this only works if there's only primitives in your data.